devito/data/data.py

```diff
@@ -155,12 +155,23 @@ def _global(self, glb_idx, decomposition):
         ret._is_distributed = any(i is not None for i in decomposition)
         return ret
 
+    def _prune_shape(self, shape):
+        # Reduce distributed MPI `Data`'s shape to that of an equivalently
+        # sliced numpy array.
+        decomposition = tuple(d for d in self._decomposition if d.size > 1)
+        retval = self.reshape(shape)
+        retval._decomposition = decomposition
+        return retval
+
     def _check_idx(func):
         """Check if __getitem__/__setitem__ may require communication across MPI ranks."""
         @wraps(func)
         def wrapper(data, *args, **kwargs):
             glb_idx = args[0]
-            if len(args) > 1 and isinstance(args[1], Data) \
+            is_gather = kwargs and isinstance(kwargs['gather_rank'], int)
```

Comment:
guarantied to have `gather_rank`? maybe `is_gather = isinstance(kwargs.get('gather_rank', None) int)`

Reply:
`0` is a valid rank so I think we need this?